tests/pos/choices.scala

```diff
@@ -0,0 +1,103 @@
+
+/** Contains a delimited contination, which can be invoked with `r`esume` */
+class Suspension[+R]:
+  def resume(): R = ???
+object Suspension:
+  def apply[R](): Suspension[R] =
+    ??? // magic, can be called only from `suspend`
+
+/** Returns `fn(s)` where `s` is the current suspension to the boundary associated
+ *  with the given label.
+ */
+def suspend[R, T](fn: Suspension[R] => T)(using Label[T]): Unit =
+  ??? // break(fn(Suspension()))
+
+/** Returns the current suspension to the boundary associated
+ *  with the given label.
+ */
+def suspend[R]()(using Label[Suspension[R]]): Unit =
+  suspend[R, Suspension[R]](identity)
+
+/** A single method iterator */
+abstract class Choices[+T]:
+  def next: Option[T]
+
+object Choices:
+  def apply[T](elems: T*): Choices[T] =
+    apply(elems.iterator)
+
+  def apply[T](it: Iterator[T]) = new Choices[T]:
+    def next = if it.hasNext then Some(it.next) else None
+end Choices
+
+/** A Label representikng a boundary to which can be returned
+ *   - None,  indicating an mepty choice
+ *   - a suspension with Option[T] result, which will be
+ *     iterated over element by element in the boundary's result.
+ */
+type CanChoose[T] = Label[None.type | Suspension[Option[T]]]
+
+/** A variable representing Choices */
+class Ref[T](choices: => Choices[T]):
+
+  /** Try all values of this variable. For each value, run the continuation.
+   *  If it yields a Some result take that as the next value of the enclosing
+   *  boundary. If it yields a Non, skip the element.
+   */
+  def each[R](using CanChoose[R]): T =
+    val cs = choices
+    suspend()
+    cs.next.getOrElse(break(None))
+
+end Ref
+
+/** A prompt to iterate a compputation `body` over all choices of
+ *  variables which it references.
+ *  @return all results in a new `Choices` iterator,
+ */
+def choices[T](body: CanChoose[T] ?=> T): Choices[T] = new Choices:
+  var stack: List[Suspension[Option[T]]] = Nil
+
+  def next: Option[T] =
+    boundary(Some(body)) match
+      case s: Some[T] => s
+      case None =>
+        if stack.isEmpty then
+          // no (more) variable choices encountered
+          None
+        else
+          // last variable's choices exhausted; use next value of previous variable
+          stack = stack.tail
+          stack.head.resume()
+      case susp: Suspension[Option[T]] =>
+        // A new Choices variable was encountered
+        stack = susp :: stack
+        stack.head.resume()
+
+@main def test: Choices[Int] =
+  val x = Ref(Choices(1, -2, -3))
+  val y = Ref(Choices("ab", "cde"))
+  choices:
+    val xx = x.each
+    xx + (
+      if xx > 0 then y.each.length * x.each
+      else y.each.length
+    )
+    /* Gives the results of the following computations:
+        1 + 2 * 1
+        1 + 2 * -2
+        1 + 2 * -3
+        1 + 3 * 1
+        1 + 3 * -2
+        1 + 3 * -3
+        -2 + 2
+        -2 + 3
+        -3 + 2
+        -3 + 3
+    */
+
```

Comment:
Here is a slightly different design of `suspend` which gives rise to a more "traditional" implementation of `choose`.

```scala
class Suspension[T, R]:
  def resume(arg: T): R = ???

def suspend[T, R](body: Suspension[T, R] ?=> R)(using Label[R]): T

trait Choice {
  def choose[A](choices: A*): A
}

// the handler
def choices[T](body: Choice ?=> T): List[T] =
  boundary [List[T]] {
    given Choice {
      def choose[A](choices: List[A]): A = suspend [A, List[T]] {
        choices.flatMap(a => resume(a))
      }
    }
    List(body)
  }

@main def test: List[Int] =
  def x = choose(1, -2, -3)
  def y = choose("ab", "cde")
  choices:
    val xx = x;
    xx + (
      if xx > 0 then y.length * x
      else y.length
    )
```

I purposefully annotated the types at `boundary` and `suspend` to highlight how the answer types look like.

Reply:
Please note how the implementation details of Choice are completely encapsulated at the definition site (`choices`). The user does not need to be aware that there are boundaries involved.

Reply:
Here is also a slightly modified version of async await with the above interface:

```scala

trait Future[T] {
  private var result: Option[T] = None
  private var waiting: ListBuffer[Runnable] = ListBuffer()

  def await()(using a: Async): T = a.await(this)
}

trait Async {
  def await[T](f: Future[T]): T
}

object Future {
  private def complete[T](f: Future[T], value: T): Unit =
    f.result = Some(value)
    f.waiting.foreach(Scheduler.schedule)
    f.waiting = ListBuffer()

  // a handler for Async
  def async(body: Async ?=> Unit): Unit =
    boundary [Unit] {
      given Async {
        def await[T](f: Future[T]): T = f.result match 
          case Some(x) = x
          case None => suspend [T, Unit] { k ?=> 
            val blocked = () => k.resume();
            f.waiting += blocked;
            f.await
          }
      }
      body
    }
}

def test(x1: Future[Int], x2: Future[Int]) =
  Future.async:
    println(x1.await + x2.await)
```

Reply:
Another indication that this slightly modified interface of `Suspension` is more "traditional" is that it is trivial to implement monadic reflection:

```scala

trait CanReflect[M[_]] {
  def reflect[R](mr: M[R]): R
}

trait Monadic[M[_]] {

  /**
   * Embedding of pure values into the monad M
   */
  def pure[A](a: A): M[A]

  /**
   * Sequencing of monadic values
   *
   * Implementations are required to implement sequencing in a stack-safe
   * way, that is they either need to implement trampolining on their own
   * or implement `sequence` as a tail recursive function.
   *
   * Actually the type X can be different for every call to f...
   * It is a type aligned sequence, but for simplicity we do not enforce this
   * here.
   */
  def sequence[X, R](init: M[X])(f: X => Either[M[X], M[R]]): M[R]

  /**
   * Helper to summon and use an instance of CanReflect[M]
   */
  def reflect[R](mr: M[R])(using r: CanReflect[M]): R = r.reflect(mr)

  /**
   * Reify a computation into a monadic value
   */
  def reify[R](prog: CanReflect[M] ?=> R): M[R] = boundary [M[R]] {
    given CanReflect[M] {
      // We could even shorten this to `suspend(mr.flatMap)` which makes the connection obvious.
      def reflect[R](mr: M[R]): R = suspend [R, M[R]] { k ?=> mr.flatMap { r => k.resume(r) } }
    }
  }
}
```
This is also a practical indication of its usefulness, since monadic reflection (we might need a less scary name) is a perfect way of switching back-and-forth between "old" monadic library implementations and new ones using control effect, as illustrated in our repo:

https://github.com/lampepfl/monadic-reflection/blob/main/core/src/main/scala/monadic/examples/FutureIO.scala

Reply:
I played a bit with these alternative encodings. I completed the futures sketch to working test case in [9538783](https://github.com/lampepfl/dotty/pull/16739/commits/9538783d9b53f05f2fe50a950199213315d6bc72). There's a slight hitch with variance, if we want to make `Future` covariant. But it can be solved with an @uncheckedVariance annotation since the usage is arguably sound. I left a comment in the test case. 

But I have a more serious doubt. How is this snipped supposed to work:
```scala
        def await[T](f: Future[T]): T = f.result match
          case Some(x) => x
          case None =>
            suspend[T, Unit]: s ?=>
              f.waiting += (v => s.resume(v))
              f.await()
```
Here, the code following `s ?=>` needs to be executed before we resume since it installs the suspension in the waiting set of the other future. But then we follow this by `f.await()`. Isn't that an infinite recursion?

Reply:
I think I have it: We just need to shift the final `await` two spaces to the left:
```scala
        def await[T](f: Future[T]): T = f.result match
          case Some(x) => x
          case None =>
            suspend[T, Unit]: s ?=>
              f.waiting += (v => s.resume(v))
            f.await
```
Then it should work. Right?

Reply:
If I understand correctly, the difference between the previous and alternative encoding amounts the following: 

In the previous encoding, we return a suspension as part of the value to the enclosing boundary. The boundary then acts on the suspension by adding it to the `waiting` set. In the new encoding, we install the suspension directly and simply return with `()` to the boundary. 

Furthermore, in the previous encoding, suspend() always was of type `Unit` whereas now it returns a value that gets passed to `resume`. At least as far as futures are concerned, that part seemed make things a bit more complicated.

Reply:
I added a middle version `futuresALT2` that reverts back to unit-valued suspends while keeping the general design ofg `futuresALT`.

Reply:
Since we resume with a value, and if I am not completely mistaken, then we don't need the `f.await` in:
```diff
suspend[T, Unit]: s ?=>
  f.waiting += (v => s.resume(v))
-f.await
```

Reply:
Yes, correct. I pushed a simplified version.